agenthub/codeact_agent/codeact_agent.py

```diff
@@ -227,13 +230,27 @@ def _get_messages(self, state: State) -> list[dict[str, str]]:
         # the latest user message is important:
         # we want to remind the agent of the environment constraints
         latest_user_message = next(
-            (m for m in reversed(messages) if m['role'] == 'user'), None
+            (
+                m
+                for m in reversed(messages)
+                if m.role == 'user'
+                and any(isinstance(c, TextContent) for c in m.content)
+            ),
+            None,
         )
 
-        # add a reminder to the prompt
+        # Get the last user text inside content
         if latest_user_message:
-            latest_user_message['content'] += (
-                f'\n\nENVIRONMENT REMINDER: You have {state.max_iterations - state.iteration} turns left to complete the task. When finished reply with <finish></finish>'
+            latest_user_message_text = next(
+                (
+                    t
+                    for t in reversed(latest_user_message.content)
+                    if isinstance(t, TextContent)
+                )
```

Comment:
Can we *concatenate* all the text from the user instead of return the last one?

Reply:
i am returning the last message to append the Reminder to it, will concatenation work here? 
```
        if latest_user_message_text:
            latest_user_message_text.text = f'{latest_user_message_text.text}\n\nENVIRONMENT REMINDER: You have {state.max_iterations - state.iteration} turns left to complete the task. When finished reply with <finish></finish>.'
```

Reply:
something like this?

```
if latest_user_message:
            latest_user_message_texts = '\n\n'.join
                ((
                    t
                    for t in reversed(latest_user_message.content)
                    if isinstance(t, TextContent)
                ))
```

Reply:
i am just trying to understand this, suppose our message looks like this:
```
[
  {
    "role": "system",
    "content": "....."
  },
  {
    "role": "user",
    "content": [
      {
        "type": "text",
        "text": "find the equation from the image"
      },
      {
        "type": "url",
        "url": "equation.png"
      },
      {
        "type": "text",
        "text": "add this equation"
      },
      {
        "type": "url",
        "url": "equation url"
      }
    ]
  }
]

```
how do you want it to be sent to the llm(since we have to add the reminder), what should the final json json look like.

Reply:
ahhh i see! looks like i misunderstand it :(

How about we add an "text" content, if the last message is not text, otherwise, we append to the last text (like what you are doing now?)

Reply:
cool , makes sense.